src/Compilers/CSharp/Portable/Binder/Binder_Flags.cs

```diff
@@ -91,11 +91,26 @@ internal Binder WithAdditionalFlagsAndContainingMemberOrLambda(BinderFlags flags
             return new BinderWithContainingMemberOrLambda(this, this.Flags | flags, containing);
         }
 
-        internal Binder WithUnsafeRegionIfNecessary(SyntaxTokenList modifiers)
+        internal Binder WithUnsafeRegionIfNecessary(SyntaxTokenList modifiers, bool isIterator = false)
         {
-            return (this.Flags.Includes(BinderFlags.UnsafeRegion) || !modifiers.Any(SyntaxKind.UnsafeKeyword))
-                ? this
-                : new Binder(this, this.Flags | BinderFlags.UnsafeRegion);
+            var insideUnsafe = this.Flags.Includes(BinderFlags.UnsafeRegion);
+            var hasUnsafeModifier = modifiers.Any(SyntaxKind.UnsafeKeyword);
+
+            if (insideUnsafe == hasUnsafeModifier)
+            {
+                return this;
+            }
+
+            if (insideUnsafe)
+            {
+                Debug.Assert(!hasUnsafeModifier);
+                return isIterator && this.Compilation.IsFeatureEnabled(MessageID.IDS_FeatureRefUnsafeInIteratorAsync)
```

Comment:
>isIterator && this.Compilation.IsFeatureEnabled(MessageID.IDS_FeatureRefUnsafeInIteratorAsync) [](http://example.com/codeflow?start=23&length=94)

The check for language version here confused me (again). Consider adding a comment here that we're not clearing `BinderFlags.UnsafeRegion` for earlier language versions to ensure that nested functions in earlier language versions inherit the `unsafe` context from the iterator container (assuming that is the reason).